### PR TITLE
resolves #111 add api to manage synonyms

### LINKED ISSUE #111: Add api to manage synonyms

### PR DESCRIPTION
**Pull request by elisaMy**

### CODE REVIEW

File: tests/api/test_agents_service.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        mocker.patch("api.agents_service.SynonymsRepository.update_synonym", return_value=True)
+        agents_service.AgentsService.get_instance().update_agent_synonym(self.agent_name, self.id, self.synonym)
+        agents_service.AgentsRepository.agent_modified.assert_called()
+
+    def test_should_not_call_agent_modified_when_update_synonym_returns_false(self, mocker):
+        mocker.patch("api.agents_service.SynonymsRepository.update_synonym", return_value=False)
+        agents_service.AgentsService.get_instance().update_agent_synonym(self.agent_name, self.id, self.synonym)
+        agents_service.AgentsRepository.agent_modified.assert_not_called()
+
+    def test_should_return_true_when_is_valid_returns_true_and_find_synonym_returns_something(self, mocker):
```
Comment (omar-mellouli) [inline tests/api/test_agents_service.py:40]: rename like:
test_should_synonym_exist_when_find_synonym_returns_something

File: tests/api/test_agents_service.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        mocker.patch("api.agents_service.ObjectId.is_valid", return_value=True)
+        mocker.patch("api.agents_service.SynonymsRepository.find_synonym", return_value="synonym")
+        assert agents_service.AgentsService.get_instance().synonym_exist(self.id) == True
+
+    def test_should_return_false_when_is_valid_returns_true_and_find_synonym_returns_none(self, mocker):
```
Comment (omar-mellouli) [inline tests/api/test_agents_service.py:45]: rename like:
test_should_synonym_not_exist_when_find_synonym_returns_none

File: tests/api/test_agents_service.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        mocker.patch("api.agents_service.ObjectId.is_valid", return_value=True)
+        mocker.patch("agents_service.SynonymsRepository.find_synonym", return_value=None)
+        assert agents_service.AgentsService.get_instance().synonym_exist(self.id) == False
+
+    def test_should_return_false_when_is_valid_returns_false_and_find_synonym_returns_something(self, mocker):
```
Comment (omar-mellouli) [inline tests/api/test_agents_service.py:50]: rename like:
test_should_synonym_not_exist_when_id_is_not_valid

**Review by omar-mellouli — CHANGES_REQUESTED** [review]

File: api/persistence/synonyms_repository.py [hunk 4]
```diff
@@ -14,6 +18,20 @@ def insert_synonyms(self, data):
             return True
         except:
             return False 
-    
+
     def delete_synonyms(self, agent_name):
```
[the remark this answers is not in the document]
Comment (elisaMy) [reply api/persistence/synonyms_repository.py]: done

File: api/persistence/synonyms_repository.py [hunk 5]
```diff
@@ -14,6 +18,20 @@ def insert_synonyms(self, data):
...
         self.synonyms_collection.delete_many({"agent_name": agent_name})
+
+    def delete_synonym_by_id(self, agent_name, id):
```
[the remark this answers is not in the document]
Comment (elisaMy) [reply api/persistence/synonyms_repository.py]: done

File: api/persistence/synonyms_repository.py [hunk 6]
```diff
@@ -14,6 +18,20 @@ def insert_synonyms(self, data):
...
+            return True
+        except:
+            return False
+
+    def update_synonym(self, agent_name, id, data):
+        try:
+            self.synonyms_collection.update_one({"agent_name": agent_name, "_id": ObjectId(id)}, {"$set": {"data": data}}, upsert = True)
+            return True
+        except:
+            return False
```
[the remark this answers is not in the document]
Comment (elisaMy) [reply api/persistence/synonyms_repository.py]: done

File: tests/api/test_agents_service.py (hunk 1 above)
Comment (elisaMy) [reply tests/api/test_agents_service.py]: done

File: tests/api/test_agents_service.py (hunk 2 above)
Comment (elisaMy) [reply tests/api/test_agents_service.py]: done

File: tests/api/test_agents_service.py (hunk 3 above)
Comment (elisaMy) [reply tests/api/test_agents_service.py]: done

File: api/routes/synonym.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+def add_synonyms(agent_name):
+    if request.get_data():
+        request_data = json.loads((request.get_data()).decode())
+        if (request_data.get("data") and isinstance(request_data.get("data"), list)):
+            AgentsService.get_instance().add_agent_synonyms(agent_name, request_data["data"])
+            return response_template(200, "Synonyms successfully added for agent {0}".format(agent_name))
+        else:
+            return response_template(400, "Should contain valid synonyms array object")
+    else:
+        return response_template(400, "A body is mandatory inside the request")
```
Comment (omar-mellouli) [inline api/routes/synonym.py:50]: Add new line

Comment (elisaMy) [reply api/routes/synonym.py]: Done

**Pull request merged**